### PR TITLE
WIP: Refactor HostedCluster AWS role API

### PR DESCRIPTION
**Pull request by ironcladlou**
Before this commit, the HostedCluster AWS role API was considered difficult to
understand while also leaking implementation details to the end user. The user
is actually responsible for creating specific required AWS policy documents and
providing the ARNs for those roles as API input. Before this change, the user
was required to supply not only the ARNs, but also the names and namespaces of
secrets the inputs are mapped to internally by HyperShift, which are
implementation details. This is a problem because not only are those constant
details irrelevant to the end user, they could change in a future HyperShift
release and thus introduce compatibility or migration problems.

This commit improves the API and implements [HOSTEDCP-153](https://issues.redhat.com//browse/HOSTEDCP-153) by refactoring the
opaque role array into a struct which requires only the ARNs for concrete roles
whose policy document requirements are documented on concrete fields. The
internal details of how these ARNs are used is no longer exposed to the end
user.

### CODE REVIEW

**Comment by relyt0925** [comment]
No concerns on API breakage from me since it's tied to aws specific fields not being utilized in production with IBM

File: api/v1alpha1/hostedcluster_types.go [hunk 1]
```diff
@@ -846,10 +837,134 @@ type AWSResourceTag struct {
...
+	//				"route53:ChangeResourceRecordSets"
+	//			],
+	//			"Resource": [
+	//				"arn:aws:route53:::PUBLIC_ZONE_ID",
+	//				"arn:aws:route53:::PRIVATE_ZONE_ID"
+	//			]
+	//		}
+	//	]
+	// }
+	IngressOIDC string `json:"ingressOidc"`
```
Comment (alvaroaleman) [inline api/v1alpha1/hostedcluster_types.go:872]: In general I really like to have this properly typed. But what will we do if the set of required roles changes across OCP versions? So far, we went with "always create everything any known OCP version needs", do we want to continue like that?

Comment (ironcladlou) [reply api/v1alpha1/hostedcluster_types.go:872]: Does the array somehow account for that? It still requires the user know magic names to correspond an ARN with its purpose, and the list contents may need to change across versions in the scenario you describe.

Is the difference thus aesthetics/ergonomics? e.g. if a new field becomes required, either the user has to start adding a new magic name to the list (in the array), or we have to add a new optional field (in the struct)?

Is there a certain example you had in mind that we can use to compare how either approach holds up?

Comment (enxebre) [reply api/v1alpha1/hostedcluster_types.go:872]: @alvaroaleman good point. Once thought is if the assumptions ever change we could let all these fields become optional and have backend validation per ocp version.

Comment (alvaroaleman) [reply api/v1alpha1/hostedcluster_types.go:872]: Yeah, makes sense, just having all fields we might need for any version and then validating seems fine, still a net improvement over what we have today

**Pull request closed without merging**